proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -4307,6 +4364,178 @@ lemma update_sched_context_rewrite:
+   apply (frule sc_inv_state_eq, simp)
+   apply (rule_tac x="(sc, s)" in bexI[rotated], simp)
+   apply (rule disjI2)
+   apply (drule use_valid[OF _ get_sc_ko'], simp)
+   apply (clarsimp simp: obj_at'_def projectKOs)
+   apply (prop_tac "obj_at' (\<lambda>k. objBits k = objBits (g (f sc))) scPtr s")
+    apply (clarsimp simp: obj_at'_def projectKOs projectKO_opt_sc)
+    apply (rule_tac x=sc in exI, clarsimp simp: projectKO_opt_sc)
+   apply (drule_tac ob1="g (f sc)" in no_failD[OF no_fail_setObject_other, rotated])
+    apply simp
+   apply clarsimp
+  apply (frule sc_inv_state_eq, simp)
+  apply (rule_tac x="(sc, s)" in bexI[rotated], simp)
+  apply (drule use_valid[OF _ get_sc_ko'], simp)
+  apply (erule disjE; clarsimp)
+   apply (clarsimp simp: obj_at'_def projectKOs)
+   apply (prop_tac "obj_at' (\<lambda>k. objBits k = objBits (f sc)) scPtr s")
+    apply (clarsimp simp: obj_at'_def projectKOs projectKO_opt_sc)
+    apply (rule_tac x=sc in exI, clarsimp simp: projectKO_opt_sc)
+   apply (drule_tac ob1="(f sc)" in no_failD[OF no_fail_setObject_other, rotated])
+    apply simp+
+
+  apply (rename_tac s'; erule disjE; clarsimp?)
+   apply (drule_tac Q2="\<lambda>s'. s' = (s\<lparr>ksPSpace := ksPSpace s(scPtr \<mapsto> injectKO (f sc))\<rparr>)"
+                 in use_valid[OF _ setObject_sc_wp])
+    apply simp+
+
+   apply (prop_tac "sc_at' scPtr (s\<lparr>ksPSpace := ksPSpace s(scPtr \<mapsto> KOSchedContext (f sc))\<rparr>)")
+    apply (clarsimp simp: obj_at'_def projectKOs objBits_simps' ps_clear_upd)
+   apply (frule_tac s="s\<lparr>ksPSpace := ksPSpace s(scPtr \<mapsto> KOSchedContext (f sc))\<rparr>"
+                 in no_failD[OF no_fail_getMiscObject(4)])
+   apply clarsimp
+
+  apply (rename_tac s')
+  apply (drule_tac Q2="\<lambda>s'. s' = (s\<lparr>ksPSpace := ksPSpace s(scPtr \<mapsto> injectKO (f sc))\<rparr>)"
+                in use_valid[OF _ setObject_sc_wp])
+   apply simp+
+
+  apply (frule sc_inv_state_eq, simp)
+  apply (drule use_valid[OF _ get_sc_ko'], simp)
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (prop_tac "obj_at' (\<lambda>k. objBits k = objBits (g (f sc))) scPtr
+                           (s\<lparr>ksPSpace := ksPSpace s(scPtr \<mapsto> KOSchedContext (f sc))\<rparr>)")
+   apply (clarsimp simp: obj_at'_def projectKOs projectKO_opt_sc)
+   apply (rule_tac x="f sc" in exI, clarsimp simp: projectKO_opt_sc)
+  apply (drule_tac ob1="g (f sc)" in no_failD[OF no_fail_setObject_other, rotated])
+   apply simp+
+  done
+
+lemmas getSchedContext_setSchedContext_decompose_decompose_ext
+  = getSchedContext_setSchedContext_decompose[where f="f x" and g="g y" for f g x y]
+lemmas getSchedContext_setSchedContext_decompose_decompose2
+  = getSchedContext_setSchedContext_decompose[where g="\<lambda>sc. g (h sc)" for g h]
+lemmas getSchedContext_setSchedContext_decompose_decompose_ext2
+  = getSchedContext_setSchedContext_decompose[where f="f x" and g="g y" for f g x y]
```

Comment:
@lsf37 : how much do we indent the `=`? I've seen four spaces in this PR, I've seen people putting the `=` under the lemmas name, etc.

Reply:
I would of course prefer it to be on the end of the previous line. Otherwise, standard indent would be 2, I think.

Reply:
The nitpicky question is "to whom does the `=`" belong to. Clearly it's under `lemmas`, but if it is under the name as well, then:

```
lemmas xxx
       = ...
```

In your version with the `=` at EOL, the same question arises, where does the `...` go? Does it belong to the name or the `lemmas`? At the moment I'm siding with it belonging with the name, not `lemmas`, but not clear cut on it.

Reply:
(of course this doesn't need changing in this commit, but more for style guide later)

Reply:
If it was a normal `lemma` statement (not `lemmas`), I think we would also continue with a 2-indent, and not indent with the name, so I'm for staying consistent with that.

Reply:
Works for me!